### PR TITLE
route: implementation of HeaderAppendAction

### LINKED ISSUE #22713: [Question] How can I insert a HTTP header only when that header does not exist?
For example, I want to add a HTTP header `foo: bar` into request when the original request does not have the HTTP header `foo`. But, if the original request have the HTTP header `foo`, envoy should do nothing header manipulation for this request.

The existing setting [HeaderValueOption](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-headervalueoption) can't resolve this issue. Please correct me if i am wrong.

CC @wbpcode

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: route: implementation of HeaderAppendAction
Additional Description:

The HeaderAppendAction was added in the #18246. But the implementation never done. This PR try to complete this feature.

Further work of #18246. To close #22713.

Risk Level: Low.
Testing: Unit Test.
Docs Changes: n/a.
Release Notes: wait.
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/router/header_parser.cc [hunk 1]
```diff
@@ -214,45 +211,59 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
...
+      // 'append' is set and ensure the 'append_action' value is equal to the default value.
+      if (header_value_option.append_action() != HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD) {
+        throw EnvoyException("Both append and append_action are set and it's not allowed");
+      }
+
+      append_action = header_value_option.append().value()
+                          ? HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD
+                          : HeaderAppendOption::OVERWRITE_IF_EXISTS_OR_ADD;
+    } else {
+      append_action = header_value_option.append_action();
```
Comment (htuch) [inline source/common/router/header_parser.cc:236]: if `append` is false, then I don't see how we can accept a default action here which is to essentially append (`APPEND_IF_EXISTS_OR_ADD`). Maybe we should only interpret `append_action` if `append` is true, in which case the semantics seem to make more sense.

Comment (wbpcode) [reply source/common/router/header_parser.cc:236]: `append` is optional bool. We will use `append_action` only when the we don't set `append`.

If `append` is set to false, the `OVERWRITE_IF_EXISTS_OR_ADD` will be used. 

So there should be no any behaviors change.

File: api/envoy/config/core/v3/base.proto [hunk 2]
```diff
@@ -358,8 +358,10 @@ message HeaderValueOption {
   // existing values. Otherwise it replaces any existing values.
   google.protobuf.BoolValue append = 2;
```
Comment (soulxu) [inline api/envoy/config/core/v3/base.proto:359]: We should be able to mark this as deprecated.

File: source/common/router/header_parser.cc [hunk 3]
```diff
@@ -214,45 +211,59 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
...
   HeaderParserPtr header_parser(new HeaderParser());
 
   for (const auto& header_value_option : headers_to_add) {
-    const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
-    HeaderFormatterPtr header_formatter = parseInternal(header_value_option.header(), append);
+    HeaderAppendAction append_action;
+
+    if (header_value_option.has_append()) {
+      // 'append' is set and ensure the 'append_action' value is equal to the default value.
+      if (header_value_option.append_action() != HeaderAppendOption::APPEND_IF_EXISTS_OR_ADD) {
```
Comment (soulxu) [inline source/common/router/header_parser.cc:228]: This means if the user explicitly set `append` as true and `append_action` as APPEND_IF_EXISTS_OR_ADD is valid config.

But if we consider to deprecated `append` field, then this case should be invalid also. We encourage the user use the new field,  if both fields are specified, it should consider as invalid even if their value not conflict.

**Review by soulxu — CHANGES_REQUESTED** [review]

File: test/common/router/header_formatter_test.cc [hunk 4]
```diff
@@ -1252,14 +1248,14 @@ match: { prefix: "/new_endpoint" }
...
-    append: true
+    append_action: APPEND_IF_EXISTS_OR_ADD
   - header:
       key: "x-client-ip-port"
       value: "%DOWNSTREAM_REMOTE_ADDRESS%"
   - header:
       key: "x-client-port"
       value: "%DOWNSTREAM_REMOTE_PORT%"
-    append: true
+    append_action: APPEND_IF_EXISTS_OR_ADD
```
Comment (htuch) [inline test/common/router/header_formatter_test.cc:1258]: There is a lot of test churn. Can you point to the tests that still provide coverage of the deprecated functionality? We should have some until the old field is retired.

File: api/envoy/config/core/v3/base.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/config/core/v3/base.proto:359]: +1

File: source/common/router/header_parser.cc (hunk 1 above)
Comment (htuch) [reply source/common/router/header_parser.cc:236]: Got it, makes sense.

File: test/common/router/header_formatter_test.cc (hunk 4 above)
Comment (wbpcode) [reply test/common/router/header_formatter_test.cc:1258]: Yeah, some more test for deprecated `append` is on the way.

File: api/envoy/config/core/v3/base.proto (hunk 2 above)
Comment (wbpcode) [reply api/envoy/config/core/v3/base.proto:359]: of course. 😸

File: source/common/router/header_parser.cc (hunk 3 above)
Comment (wbpcode) [reply source/common/router/header_parser.cc:228]: I think there is no way to distinguish a explicit APPEND_IF_EXISTS_OR_ADD or default APPEND_IF_EXISTS_OR_ADD. 🤔

Comment (soulxu) [reply source/common/router/header_parser.cc:228]: emm...I thought there should be an interferce `has_append_action`, seems not.

**Comment by wbpcode** [comment]
It's ready for a formal review for now. cc @htuch @soulxu 😸

File: source/common/router/header_parser.h [hunk 5]
```diff
@@ -38,16 +41,16 @@ class HeaderParser : public Http::HeaderEvaluator {
    */
   static HeaderParserPtr
   configure(const Protobuf::RepeatedPtrField<envoy::config::core::v3::HeaderValue>& headers_to_add,
-            bool append);
```
Comment (soulxu) [inline source/common/router/header_parser.h:41]: Forget to remove the comment for `append` at https://github.com/envoyproxy/envoy/pull/22723/files#diff-b5669de72ffdab01ee2e54f54e9cb97f9d6f7bf4e053fb31c6af1ee56f2cf316R39

File: source/common/router/header_parser.cc [hunk 6]
```diff
@@ -214,45 +211,59 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
...
 
   for (const auto& header_value_option : headers_to_add) {
-    const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
-    HeaderFormatterPtr header_formatter = parseInternal(header_value_option.header(), append);
+    HeaderAppendAction append_action;
+
+    if (header_value_option.has_append()) {
+      // 'append' is set and ensure the 'append_action' value is equal to the default value.
+      if (header_value_option.append_action() != HeaderValueOption::APPEND_IF_EXISTS_OR_ADD) {
+        throw EnvoyException("Both append and append_action are set and it's not allowed");
```
Comment (soulxu) [inline source/common/router/header_parser.cc:229]: seems like lack unittest for this case.

**Review by soulxu — CHANGES_REQUESTED** [review]
LGTM, two small comments

File: source/common/router/header_parser.h (hunk 5 above)
Comment (wbpcode) [reply source/common/router/header_parser.h:41]: right. Thanks.

**Review by soulxu — APPROVED** [review]
LGTM, thanks!

File: source/common/router/header_parser.cc [hunk 7]
```diff
@@ -285,10 +296,20 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
...
+      case HeaderValueOption::ADD_IF_ABSENT:
+        if (auto header_entry = headers.get(key); header_entry.empty()) {
+          headers.addReferenceKey(key, value);
+        }
+        break;
+      case HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD:
         headers.setReferenceKey(key, value);
+        break;
+      default:
+        PANIC_DUE_TO_CORRUPT_ENUM;
```
Comment (htuch) [inline source/common/router/header_parser.cc:312]: Looking back at https://github.com/envoyproxy/envoy/pull/19204, I think we want `PANIC_ON_PROTO_ENUM_SENTINEL_VALUES`.

File: test/common/router/header_formatter_test.cc [hunk 8]
```diff
@@ -1639,6 +1640,132 @@ request_headers_to_remove: ["x-foo-header"]
...
+  {
+    Http::TestResponseHeaderMapImpl header_map{{"x-per-header", "exist-per"}};
+    resp_header_parser->evaluateHeaders(header_map, stream_info);
+    EXPECT_EQ("foo", header_map.get_("x-foo-header"));
+    EXPECT_EQ("bar", header_map.get_("x-bar-header"));
+    EXPECT_EQ("exist-per", header_map.get_("x-per-header"));
+  }
+}
+
+TEST(HeaderParserTest, DEPRECATED_FEATURE_TEST(EvaluateRequestHeadersAddWithDeprecatedAppend)) {
```
Comment (htuch) [inline test/common/router/header_formatter_test.cc:1701]: Are these the only tests for the deprecated append?

File: api/envoy/config/core/v3/base.proto [hunk 9]
```diff
@@ -334,7 +334,7 @@ message HeaderValueOption {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.api.v2.core.HeaderValueOption";
 
-  // [#not-implemented-hide:] Describes the supported actions types for header append action.
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply api/envoy/config/core/v3/base.proto:337]: What do you mean by the `version history`? This `HeaderAppendAction` has no previous version.

File: source/common/router/header_parser.cc (hunk 7 above)
Comment (wbpcode) [reply source/common/router/header_parser.cc:312]: get it.

**Comment by wbpcode** [comment]
Because some special behavior in the `ext_authz`/`ext_proc` (#22845), most update in the ext_authz & ext_proc has been reverted.

File: api/envoy/config/core/v3/base.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (htuch) [reply api/envoy/config/core/v3/base.proto:337]: https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.24/v1.24.0

[the remark this answers is not in the document]
Comment (wbpcode) [reply api/envoy/config/core/v3/base.proto:337]: I see. The release note. Get it. Thanks.

File: source/common/router/header_parser.cc [hunk 10]
```diff
@@ -55,7 +54,7 @@ HeaderFormatterPtr parseInternal(const envoy::config::core::v3::HeaderValue& hea
 
   absl::string_view format(header_value.value());
   if (format.empty()) {
-    return std::make_unique<PlainHeaderFormatter>("", append);
```
Comment (htuch) [inline source/common/router/header_parser.cc:58]: For my education, why did we ever have append here?

Comment (htuch) [reply source/common/router/header_parser.cc:58]: I see actually, you now propagate this independent of the header formatter, seems cleaner.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**